crates/core/c8y_api/src/smartrest/message.rs

```diff
@@ -21,9 +27,47 @@ pub fn get_smartrest_template_id(payload: &str) -> String {
     payload.split(',').next().unwrap().to_string()
 }
 
+/// Sanitize the input to be SmartREST compatible.
+/// If the input contains invalid UTF-8, it returns an empty String.
+/// - Remove all control characters except for `\n`, `\t`, `\r`.
+/// - Double quote is escaped as `\"`.
+/// - Strip the input according to `max_size`.
+pub fn sanitize_for_smartrest(input: Vec<u8>, max_size: usize) -> String {
+    String::from_utf8(input)
+        .unwrap_or_else(|err| {
+            error!("The input contains invalid UTF-8: {err}");
+            String::default()
+        })
+        .chars()
+        .filter(|&c| c == '\r' || c == '\n' || c == '\t' || !c.is_control())
+        .collect::<String>()
+        .replace('"', "\"\"")
```

Comment:
This intermediate `collect` could be avoid by using a `flat_map` instead of the `replace`.